torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,175 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+// TODO: make this thread_local + add mutexes
+std::unordered_map<std::string, ncclComm_t*> _communicators;
+
+PyObject * THCPModule_nccl_destroy(PyObject *args) {
+  for(auto it: _communicators) {
+    ncclCommDestroy(*(it.second));
+  }
+  Py_RETURN_NONE;
+}
+
+static ncclComm_t* _get_communicator(std::vector<at::Tensor>& inputs, int ndevices) {
+  std::stringstream hash;
+  std::vector<int> devs;
+  for (int i = 0; i < ndevices; i++) {
+    int dev = inputs[i].get_device();
+    hash <<  dev << ",";
+    devs.push_back(dev);
+  }
+  if (_communicators.find(hash.str()) == _communicators.end()) {
+    ncclComm_t *comms = (ncclComm_t*) malloc(sizeof(ncclComm_t) * ndevices);
+    if (comms == NULL) throw std::runtime_error("failed to allocate ncclComm_t*");
+    CHECK(ncclCommInitAll(comms, ndevices, &devs[0]));
+    _communicators[hash.str()] = comms;
+  }
+  return _communicators[hash.str()];
+}
+
+static void _check_inputs(std::vector<at::Tensor> &inputs, std::vector<at::Tensor> &outputs, int size_multiplier) {
+  // len(inputs) == len(outputs)
+  size_t len = inputs.size();
+
+  if (len <= 0) {
+    throw std::runtime_error("input sequence can't be empty");
+  }
+
+  if (len != outputs.size()) {
+    std::stringstream err;
+    err << "inputs and outputs sequences have to be of the same length, but got input of length " << len << " and output of length " << outputs.size();
+    throw std::runtime_error(err.str());
+  }
+
+  std::unordered_set<int> devices;
+  int64_t numel = inputs[0].numel();
+  auto type = inputs[0].type().ID();
+
+  for (size_t i = 0; i < len; i++) {
+    auto input = inputs[i];
+    auto output = outputs[i];
+
+    if (!(input.type().isCuda() && !input.type().isSparse()
+	  && output.type().isCuda()  && !output.type().isSparse())) {
+      throw std::runtime_error("input and output elements have to be cuda Tensors");
+    }
+
+    if (type != input.type().ID() || type != output.type().ID()) {
+      throw std::runtime_error("all inputs and outputs must be of the same Tensor type");
+    }
+
+    if (! (input.is_contiguous() && output.is_contiguous())) {
```

Comment:
This `!` is really easy to miss. Maybe write it as an or of two negated contiguous checks?